refactor: Clean up and simplify imageupload

sarthaksarthak9 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:

After this PR:

- Remove `processingCompleteFunc` type and inline variables where applicable  
- Unexport constants that are not needed publicly and relocate private ones closer to usage  
- Mark mutually exclusive flags in Cobra instead of relying on custom validation logic  
- Drop support for deprecated invocation patterns  
- Make default instancetype/preference kind optional, defaulting to cluster-wide  
- Refactor `run` method to reduce cyclomatic complexity using early returns  
- Ensure HTTP request body is closed in `ConstructUploadProxyPathAsync`  
- Remove stray comment in `waitDvUploadScheduled`  
- Split `imageupload.go` into multiple files for better maintainability  
<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #14086

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

ksimon1 commented on pkg/virtctl/imageupload/imageupload.go:
```diff
@@ -23,155 +23,58 @@ import (
-	cmd.Flags().StringVar(&c.size, "size", "", "The size of the DataVolume to create (ex. 10Gi, 500Mi).")
-	cmd.Flags().StringVar(&c.storageClass, "storage-class", "", "The storage class for the PVC.")
-	cmd.Flags().StringVar(&c.accessMode, "access-mode", "", "The access mode for the PVC.")
-	cmd.Flags().BoolVar(&c.blockVolume, "block-volume", false, "Create a PVC with VolumeMode=Block (default is the storageProfile default. for archive upload default is filesystem).")
```
This parameter is not needed anymore?

ksimon1 commented on pkg/virtctl/imageupload/pvc.go:
```diff
@@ -0,0 +1,121 @@
+func (c *command) getAndValidateUploadPVC() (*v1.PersistentVolumeClaim, error) {
+	pvc, err := c.client.CoreV1().PersistentVolumeClaims(c.namespace).Get(context.Background(), c.name, metav1.GetOptions{})
+	if err != nil {
+		c.cmd.Printf("PVC %s/%s not found \n", c.namespace, c.name)
```
This is not accurate, since IsNotFound errro can be one of multiple error types  the Get function returns

ksimon1 commented on pkg/virtctl/imageupload/datavolume.go:
```diff
@@ -0,0 +1,95 @@
+		// When the PVC exists but the DV doesn't, there are two possible outcomes:
+		if k8serrors.IsNotFound(err) {
+			// 1. The DV was already garbage-collected. The PVC was created and populated by CDI as expected.
+			if dvGarbageCollected := pvc.Annotations[deleteAfterCompletionAnnotation] == "true" &&
```
Is this code still required, since DV garbage collection was deprecated?

ksimon1 commented on pkg/virtctl/imageupload/pvc.go:
```diff
@@ -0,0 +1,121 @@
+	//   a. shouldExist is true
+	//   b. shouldExist is false AND the upload annotation exists
+
+	_, isUploadPVC := pvc.Annotations[uploadRequestAnnotation]
```
move this to line 106?

ksimon1 commented on pkg/virtctl/imageupload/command.go:
```diff
@@ -0,0 +1,239 @@
+			return nil, err
+		}
+
+		if !c.noCreate && len(c.size) == 0 {
```
wouldn't be better to add this check in c.getAndValidateUploadPVC() function?

ksimon1 commented on pkg/virtctl/imageupload/pvc.go:
```diff
@@ -0,0 +1,121 @@
+	pvc := storagetypes.RenderPVC(&quantity, c.name, c.namespace, c.storageClass, c.accessMode, c.volumeMode == "block")
+	if c.volumeMode == "filesystem" {
+		volMode := v1.PersistentVolumeFilesystem
+		pvc.Spec.VolumeMode = &volMode
```
use pvc.Spec.VolumeMode = pointer.P(v1.PersistentVolumeFilesystem)

ksimon1 commented on pkg/virtctl/imageupload/pvc.go:
```diff
@@ -0,0 +1,121 @@
+	pvc.ObjectMeta.Annotations = annotations
+	c.setDefaultInstancetypeLabels(&pvc.ObjectMeta)
+
+	pvc, err = c.client.CoreV1().PersistentVolumeClaims(c.namespace).Create(context.Background(), pvc, metav1.CreateOptions{})
```
```
return c.client.CoreV1().PersistentVolumeClaims(c.namespace).Create(context.Background(), pvc, metav1.CreateOptions{})
```

ksimon1 commented on pkg/virtctl/imageupload/pvc.go:
```diff
@@ -0,0 +1,121 @@
+
+func (c *command) ensurePVCSupportsUpload(pvc *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	var err error
+	_, hasAnnotation := pvc.Annotations[uploadRequestAnnotation]
```
Why are you checking here if annotations contains string and then inside you are checking if annotations are nil?

ksimon1 commented on pkg/virtctl/imageupload/util.go:
```diff
@@ -0,0 +1,145 @@
+
+	switch c.volumeMode {
+	case "block":
+		volMode := v1.PersistentVolumeBlock
```
inline this:
spec.VolumeMode = pointer.P(spec.VolumeMode)
The same for filesystem mode

sarthaksarthak9 commented on pkg/virtctl/imageupload/wait.go:
```diff
@@ -0,0 +1,115 @@
+	return err
+}
+
+func waitUploadProcessingComplete(client kubernetes.Interface, cmd *cobra.Command, namespace, name string, interval, timeout time.Duration) error {
```
Yes, It is used in imageupload.go.

sarthaksarthak9 commented:
@ksimon1 Thank you for your thorough review of the PR. I’ve addressed all your comments in this revision. Please feel free to let me know if there’s anything else to refine—appreciating your insights!

ksimon1 commented on pkg/virtctl/imageupload/util.go:
```diff
@@ -0,0 +1,153 @@
+
+	// Retrieve events filtered by involved object
+	eventList, err := c.client.CoreV1().Events(c.namespace).List(context.Background(), metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("involvedObject.name=%s", c.name),
```
Please use `fields.OneTermEqualSelector` instead of `fmt.Sprintf`

ksimon1 commented:
@0xFelix would you please check this PR?

0xFelix reviewed (commented):
@sarthaksarthak9 Thank you for your work! Can you please split this into multiple commits? Having all changes in a single commit makes it really hard to review this.

sarthaksarthak9 commented:
@0xFelix Thanks for the feedback! I totally agree—splitting this into multiple commits will make it easier to review. I’ll break it down into logical chunks

sarthaksarthak9 commented:
> @sarthaksarthak9 Thank you for your work! Can you please split this into multiple commits? Having all changes in a single commit makes it really hard to review this.

@0xFelix Thanks for the feedback! I totally agree—splitting this into multiple commits will make it easier to review.
